### ISSUE TITLE
Implement oil management

### ISSUE THREAD

**Issue by gilbertohasnofb** [HHS81/c182s#125]
Implement oil management, create a dialog for checking oil level and create a hotspot around the place the oil tank cap in the aircraft nose (see c172p for an implementation of this idea)

**Comment by gilbertohasnofb** [comment]
I wanted to start working on this, but here are a couple of questions before:
- where exactly is the oil tank on the 182S? Is this the cap?:
![](http://cdn2.planeandpilotmag.com/stories/2008/june/ntsb-IMG_1790.jpg)
- if it is, I think we should add it to the liveries, or better yet to a heightmap like we did for the c172p
- once we know where the oil tank cap is, to implement this, I'd need help from someone to create a hotspot just like @wlbragg did for the c172p

**Comment by gilbertohasnofb** [comment]
Info for myself: oil tank can take up to 9 quarts, critical value is 4 quarts

**Comment by HHS81** [comment]
The oil cap is on the right fuselage in front of the door. It is drawn on the texture and included by the normalmap I think.

I can add a hotspot but not today. 

Please consider a use of different types of oil. This can be simulated by jsbsim and Cannstatt bei propably implemented by me. Thanks to Ronald again!

**Comment by HHS81** [comment]
To make sure: Oil cap: not on the Top of the engine cowling but on the side.

**Comment by HHS81** [comment]
Ronald = Ron!!! F*** automatic correction on my Smartphone

**Comment by gilbertohasnofb** [comment]
> The oil cap is on the right fuselage in front of the door. It is drawn on the texture and included by the normalmap I think.

Oh, I see it now! Perfect :+1: 

> I can add a hotspot but not today. 

No rush

> Please consider a use of different types of oil. This can be simulated by jsbsim and Cannstatt bei propably implemented by me. 

Hmm, I don't know how that would work then. In the c172p, I implemented it using a Nasal function. Basically if the engine is on, the oil level goes down (depending also on RPM). If oil level very low, engine coughs, if critical then it fails or can't be started. 

If you want something fancier than that than I might not be the right person to work on this issue. But if you don't mind me asking, do you really think we need different types of oil? I personally don't see the need.

**Comment by HHS81** [comment]
> Hmm, I don't know how that would work then. In the c172p, I implemented it using a Nasal function. Basically if the engine is on, the oil level goes down (depending also on RPM). If oil level very low, engine coughs, if critical then it fails or can't be started.

Nothing wrong with (beside using nasal here - nasal really needed for that?)
But each types of oil has different viscosity and decreases engine power more or less. And I think it has an influence on the oil level as well. So user can decide which type of oil they want to use and which influence they want to expect (less influence on power but faster loss, or the way around....)

**Comment by HHS81** [comment]
Oh, and of course the most important thing: how temperature act on it. A less warm oil has more vicsosity --> more friction in the engine --> less power

**Comment by HHS81** [comment]
I was wrong: Cessna 182T
http://www.highlandstoday.com/storyimage/HI/20140503/ARTICLE/140509970/AR/0/AR-140509970.jpg

But not sure if true for S-Version

**Comment by HHS81** [comment]
yes: http://cdn.avweb.com/media/newspics//196730_add_oil_cessna.jpg

**Comment by HHS81** [comment]
even better visible:

**Comment by HHS81** [comment]
needs some work now from me

**Comment by gilbertohasnofb** [comment]
Heiko, in that case I think I won't be able to implement this myself, sorry :cry:

**Comment by HHS81** [comment]
You can implement the oil management, I would add the additional features on top of it. At least the influence of temperature is something I would like to have, but depends on the engine simulation, which needs improvment as you know.

**Comment by gilbertohasnofb** [comment]
The thing is I am not familiar at all with how to do this without Nasal. I will take a look on what would be required to do it using XML, and I will report back, but really feel free to grab this issue if you feel like, I think you are much more knowledgeable for this than me :smile:

**Comment by hbeni** [comment]
Also, when implementing this, we should implement external heater feature and initialize the oil temp to environment temp. As i read in the FGFS forums, jsbsim cant do this so we need a custom system for all that. I think to simulate the oil tank we might use another tank with proper density setting for oil, this way we can use the external flow rate to simulate rpm dependant oil consuming easily.

**Comment by hbeni** [comment]
Oil pressure needs also to be simulated.

From POH 4-21:
![image](https://user-images.githubusercontent.com/13608602/31579196-4088ddce-b131-11e7-84c4-8e83fca6d307.png)

**Comment by HHS81** [comment]
In the Moment you need to throttle up to see a sufficient reaction on oil temperature and pressure.

**Comment by HHS81** [comment]
Some information about jsbsim oil simulation 
http://wiki.flightgear.org/JSBSim_Engines

**Comment by HHS81** [comment]
And from the p51d: an Implementation of oil pressure and temp simulation 
https://sourceforge.net/p/flightgear/fgaddon/HEAD/tree/trunk/Aircraft/p51d/Systems/propulsion.xml

**Comment by hbeni** [comment]
I found the following information on oil specifications from [Shell](http://www.shell.com/business-customers/aviation/aeroshell/knowledge-centre/the-aeroshell-book/_jcr_content/par/textimage_1433441235.stream/1445042093501/281ec318e13a1739d0a798f601b7330a514cc1f8d08258b1157216fa806075c6/aeroshell-book-3peo.pdf):
![image](https://user-images.githubusercontent.com/13608602/31912515-71c3e874-b844-11e7-98b4-27e9bc0ff9ba.png)

Especially the viscosity and density vlues may come in handy.

**Comment by hbeni** [comment]
Another link for consumption rates (they depend on hours the oil is in service): http://ww.t-craft.org/Reference/Aircraft.Oil.Usage.pdf

**Comment by hbeni** [comment]
Another link for consumption rates (they depend on hours the oil is in service): http://www.t-craft.org/Reference/Aircraft.Oil.Usage.pdf

**Comment by hbeni** [comment]
(sorry closed this by accident)

**Comment by hbeni** [comment]
Normal operation is 8 quarts, above that higher consumption occurs.
Less than 4 quarts is critical.
![image](https://user-images.githubusercontent.com/13608602/33093209-a98c926e-cefc-11e7-9a4b-e48c61361c4d.png)

**Comment by hbeni** [comment]
Added a branch for implementing this: https://github.com/HHS81/c182s/tree/OilManagement
- [x] Basic oil consumption and management
- [x] Simulate influence of oil service time to oil consumption
- [x] Simulate higher consumption when oil tank is filled to more than 8 quarts
- [ ] update engine definition to include new FG 3.5 properties for better engine simulation
  - [ ] also correct the spark plugs value (magneto check drop should not exceed 150rpm, which it currently does)
- [ ] update gauges to be influenced by the temperature and pressure factors
- [ ] Simulate power loss when oil is still cold
- [ ] update liveries to include oil cap

**Comment by hbeni** [comment]
@gilbertohasnofb @HHS81 
The oil pressure and temperature settings need tweaking. I have no idea where to get sane values from... and also have no real life experience (to be tweaked in the [Engines/engIO540AB1A5.xml](https://github.com/HHS81/c182s/blob/OilManagement/Engines/engIO540AB1A5.xml) file)

    <oil-pressure-relief-valve-psi>   60    </oil-pressure-relief-valve-psi>
    <oil-pressure-rpm-max>          1940    </oil-pressure-rpm-max>
    <design-oil-temp-degK>           358    </design-oil-temp-degK>
    <oil-viscosity-index>              0.94 </oil-viscosity-index>

Otherwise, it looks promising!

**Comment by hbeni** [comment]
I got those numbers from a [lycoming manual](https://www.lycoming.com/sites/default/files/O%20%26%20IO-540%20Oper%20Manual%2060297-10.pdf) (p. 3-9):
![image](https://user-images.githubusercontent.com/13608602/33175016-8d9d8202-d05a-11e7-9e20-54f22c0b9c7c.png)

and (p. 3-13):
![image](https://user-images.githubusercontent.com/13608602/33175190-29b18f3a-d05b-11e7-9cbe-4d0c98d6fa1f.png)


Also, in the manual i found that priming should be done with 1/4 throttle travel open, not inch (as is currently implemented) - i will fix that.

**Comment by hbeni** [comment]
Updated the livery now to include the oil cap.
@gilbertohasnofb I also tried to add it to the normalmap, however it shows no reflection effects - what am i doing wrong?
Once the reflection is there, i can more easily update the other liveries.

**Comment by gilbertohasnofb** [comment]
@hbeni If you are modifying any of the textures in this project, please do so using the original xcf files instead of just working over the png files. We have a separate repository for those files, see https://github.com/HHS81/c182-Textures Once you finish with a xcf file and export a png, then you copy that png to this project here. Makes sense?

@HHS81 would you mind giving write access to hbeni in that repository?

> I also tried to add it to the normalmap, however it shows no reflection effects - what am i doing wrong?

I would need to see the png file, is this available in some branch?

**Comment by hbeni** [comment]
@gilbertohasnofb Thank you for the hint, i would happily work with the xcf-files. In case of my custom-callsign-livery in blue (the one of the screenshot) it is an inkscape svg-file. @HHS81 I would happily delete it from this repository and add it to the other; but what is the reason to have two separate repos for this? Would't it be more convinient for other developers to have all in the dev/ directory?

**Comment by hbeni** [comment]
@gilbertohasnofb The modified normalmap is here: https://github.com/HHS81/c182s/commit/e2ccbf7c8b6f377d493386e130611b7cdd193e8e (OilManagement branch)

**Comment by hbeni** [comment]
Just tested; the oil seems a little too cold in low power cruise flight, i had about 100° where it should be at about 150°. Will check values again.
![image](https://user-images.githubusercontent.com/13608602/33189849-1d6b82de-d0a7-11e7-9b07-e6fea58edb71.png)

**Comment by gilbertohasnofb** [comment]
> but what is the reason to have two separate repos for this? Would't it be more convinient for other developers to have all in the dev/ directory?

xcf files are huge, in particular if you use many layers like I normally do (e.g. one single xcf file for a c182s livery is around 50-60 Mb). Add to that the fact that git preserves history of files and you have a huge amount of data (our texture repo is already over 800Mb in size). There is no need to bloat the main repository with those and is common practise to have a separate repository for that (the c172p does the same, for instance).

> The modified normalmap is here:

Did you also try modifying the file `/Models/Effects/Normalmap.png`? I believe that's the file that gives the normal map effect. Also notice that you have a strange artifact (looking like an infinity symbol) above the aircraft nose in your version of the `NM.png` file:

![](https://raw.githubusercontent.com/HHS81/c182s/e2ccbf7c8b6f377d493386e130611b7cdd193e8e/Models/Effects/NM.png)

Is that intentional?

**Comment by hbeni** [comment]
yes, the artifact was a (failed) test. It is already reverted in the repo.

Am 2017-11-24 0:52, schrieb Gilberto Agostinho:

**Comment by gilbertohasnofb** [comment]
:+1:

**Comment by hbeni** [comment]
I will try to investigate the Normalmap.png and report back.
Is this file also in the other repo? (i didn't find it there)

**Comment by gilbertohasnofb** [comment]
> I will try to investigate the Normalmap.png and report back.

Great! I believe that will do the trick, but if you have any issues please let me know. 

> Is this file also in the other repo? (i didn't find it there)

No, I don't think so. We don't have all textures over there, we only add xcf files in order to preserve the layers. In the case of normalmaps, I don't think this is a big deal, but when dealing with liveries or complex panel textures it's always good to keep the xcf with as many layers as possible in order to revert/modify/add stuff in the future in a non-destructive way.

**Comment by hbeni** [comment]
> Great! I believe that will do the trick, but if you have any issues please let me know.
Yeah, how do i edit the normalmap image? do you just use gimp or is there a smarter alternative?

**Comment by gilbertohasnofb** [comment]
Yes, use GIMP. There are plugins that can help you with that, as generating a normal map by hand is not trivial at all. E.g., this is what the Normalmap plugin does; from this:

![image](https://user-images.githubusercontent.com/5700795/33190679-2a22a908-d0a7-11e7-8eb2-f5b91cf781ce.png)

into this:

![image](https://user-images.githubusercontent.com/5700795/33190674-23c79578-d0a7-11e7-9883-50a255f3fbab.png)

So basically the brightness of the black and white image is converted into height.

I am not sure this plugin is available on default GIMP or if I downloaded it from somewhere else, could you please check if you have `Filters > Map > Normalmap...`. If not, then I will take a look where did I download it from.

**Comment by hbeni** [comment]
@gilbertohasnofb i guess, the NM image i edited first (the one with the artifact) is the source image i need to edit and then convert?

**Comment by gilbertohasnofb** [comment]
> i guess, the NM image i edited first (the one with the artifact) is the source image i need to edit and then convert?

That's my guess as well, though I don't know if that map is used by anything else (I believe these maps were created by @HHS81 ). Did you check if you have the Normalmap plugin in your GIMP?

**Comment by hbeni** [comment]
@gilbertohasnofb i don't have that plugin unfortunaltely.
Is it this one: https://code.google.com/archive/p/gimp-normalmap/downloads

**Comment by gilbertohasnofb** [comment]
> i don't have that plugin unfortunaltely.
> Is it this one: https://code.google.com/archive/p/gimp-normalmap/downloads

Yeah, I thought it wasn't something that came originally with GIMP, but I can't find it on my plugins and script folders. This is what I have:

![screenshot from 2017-11-24 14 27 01](https://user-images.githubusercontent.com/5700795/33214402-b2b17250-d123-11e7-92b0-079f20bab9d4.png)

**Comment by hbeni** [comment]
@gilbertohasnofb Thanks it works! Exporting with the default values and Y-reversed yields good results.
It would have been nice, if we had a higher res image tough.
![image](https://user-images.githubusercontent.com/13608602/33219773-8e55c278-d144-11e7-90e1-255efade955b.png)

**Comment by gilbertohasnofb** [comment]
@hbeni That looks great, thanks for adding this!

Just a small comment about normal maps (and this is not a critique, just a comment about how to work with it). Please notice that having only a "contour" vs filling the inside contour will create different effects in the normal map:

![image](https://user-images.githubusercontent.com/5700795/33220663-f3038fb4-d141-11e7-8b3e-6c7aa55ed456.png)

produces (with reversed Y):

![image](https://user-images.githubusercontent.com/5700795/33220691-20ace38e-d142-11e7-8680-775bc3ab86f6.png)


So if you want something to be slightly elevated, make the whole thing brighter than the surrounding, else you get the wall or valley effect. But concerning the oil cap, it's really up to you, IMO it's looking quite good already.

Cheers!

**Comment by hbeni** [comment]
The Oil cap i have seen i the video is the same level as the fuselage, there is just this little crack; the same is true for the knob.
it looks correct in my opinion :)

Thank you for your feedback, i already noticed it while playing around (i.e.t he current look is intentional)

I think we are done with this and i will file the PR

**Comment by gilbertohasnofb** [comment]
> it looks correct in my opinion :)
> 
> Thank you for your feedback, i already noticed it while playing around (i.e.t he current look is intentional)

Great! I also think it looks good, I just wanted to let you know about it in case you were not familiar with. I really enjoy making normal maps, the effects can really bring a model to life!

**Comment by hbeni** [comment]
Thanks for all your help. I am learning a LOT since i started here!

**Comment by gilbertohasnofb** [comment]
My pleasure!

Also, here is something I wanted to ask you: would you also be interested in contributing to the c172p project? We have been looking for new developers and contributors over there as most of us are too busy at the moment with other things and the repo has stalled a bit. If that would be in your interest, let me know; given your excellent work here, you would be more than welcomed over there!

**Comment by HHS81** [comment]
@hbeni 
A normalmap is greyscale texture converted into a so called normal map. The greyscale texture converts bight areas (=white) into raised areas and dark (=black) areas into low areas. So 50% grey is level. 

It is not recommended to alter the converted normalmap, but the the source image, and then convert it again. 

The normalmap converter for Gimp has also the ability that raised points of the normalmap can get also transparency defined in the alpha layer. Transparency  controls the specular strength and can be used to make rivets and other highlighted areas more visible. You can also use it to get the surface an uneven, "dirty" touch like fingerprints on a shiny surface, which are only visible at certain light angles.

**Comment by hbeni** [comment]
@gilbertohasnofb Thank you for the honor. If i get ever complete with the c182; i may contribute also to the 172 :) Currently however i think it is better to focus on one plane only. You are very welcome to steal code/stuff from here, i always try to make things as generic as i can.

**Comment by gilbertohasnofb** [comment]
@hbeni No problems, but if you ever feel like contributing to that project please drop me a line and I would happily give you write access to it, as you are doing some excellent work here!

**Comment by hbeni** [comment]
Fixed by merge of  pr #187
